### PR TITLE
[UNR-5437] Allocate entityid during begin play

### PR DESCRIPTION
**Pull request by Ajanth-Sutharsan**
#### Description
Allocate entity ids when they're first spawned and during begin play. Linked issue: https://improbableio.atlassian.net/browse/UNR-5437

#### Tests
Tested locally

### CODE REVIEW

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp [hunk 1]
```diff
@@ -3031,7 +3038,7 @@ void USpatialNetDriver::RefreshActorDormancy(AActor* Actor, bool bMakeDormant)
 	const bool bHasAuthority = HasServerAuthority(EntityId);
 	if (bHasAuthority == false)
 	{
-		UE_LOG(LogSpatialOSNetDriver, Warning, TEXT("Unable to flush dormancy on actor (%s) without authority"), *Actor->GetName());
+		UE_LOG(LogSpatialOSNetDriver, Verbose, TEXT("Unable to flush dormancy on actor (%s) without authority"), *Actor->GetName());
```
Comment (simonsarginson) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp:3041]: Why this change?

File: SpatialGDK/Source/SpatialGDK/Public/Interop/GlobalStateManager.h [hunk 2]
```diff
@@ -66,6 +66,7 @@ class SPATIALGDK_API UGlobalStateManager : public UObject
 
 	// Startup Actor Manager Component
 	bool bCanSpawnWithAuthority;
+	bool bGSMReadyForPlay;
```
Comment (simonsarginson) [inline SpatialGDK/Source/SpatialGDK/Public/Interop/GlobalStateManager.h:69]: This is unused?

Comment (Ajanth-Sutharsan) [reply SpatialGDK/Source/SpatialGDK/Public/Interop/GlobalStateManager.h]: Removed

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp (hunk 1 above)
Comment (Ajanth-Sutharsan) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp:3080]: Since the entity id is now allocated, it doesn't go into the previous if statement, so I changed this one from warning to verbose just like the previous one

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp [hunk 3]
```diff
@@ -775,9 +775,17 @@ void USpatialNetDriver::CleanUpServerConnectionForPC(APlayerController* PC)
 		   TEXT("While trying to clean up a PlayerController, its client connection was not found and thus cleanup was not performed"));
 }
 
-void USpatialNetDriver::OnActorSpawned(AActor* Actor) const
+void USpatialNetDriver::OnActorSpawned(AActor* Actor)
```
Comment (MatthewSandfordImprobable) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp:778]: Just wondering why the loss of "const". Does PackageMap->TryResolveObjectAsEntity somehow modify something on the SpatialNetDriver?

Comment (Ajanth-Sutharsan) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp]: Ah no, the loss of const was because a previous revision did modify the netdriver, I'll add it

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp (hunk 1 above)
Comment (simonsarginson) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp:3080]: I am not sure what you mean by 'Since the entity id is now allocated'?

Comment (simonsarginson) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp:3080]: @m-samiec you wrote this code, can you share why this originally a warning?

**Review by MatthewSandfordImprobable — APPROVED** [review]

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp (hunk 1 above)
Comment (m-samiec) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp:3080]: Running this code without authority would imply that a non-auth server is attempting to modify the entity. Which suggests an error further up the chain.

The two checks are (1) do we have an entity id, and (2) do I have auth over that entity, I can see with this PR the gap between 1 and 2 has grown, so this warning is more likely to occur, however I'd like to know what the test scenario is that results in the warning as it would suggest we're modifying dormancy before we should.

If the call to modify dormancy is valid, the fix seems to be that we add an additional check between (1) and (2) which is "Do I have the entity in my view?". If not, we can return with a verbose error like we did for (1), and keep (2) as a warning.

Comment (m-samiec) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp:3080]: This would do it `Connection->GetCoordinator().GetView().HasEntity(EntityId)`

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp [hunk 4]
```diff
@@ -3040,6 +3048,20 @@ void USpatialNetDriver::RefreshActorDormancy(AActor* Actor, bool bMakeDormant)
...
+		if (ActorChannel == nullptr)
+		{
+			return;
+		}
+
+		if (ActorChannel->bCreatedEntity)
+		{
+			ActorSystem->AddEntityToRefreshDormancy(EntityId, bMakeDormant);
+			return;
+		}
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp:3063]: need to add an else here, and error as it's not valid to continue if we don't have the entity checked out

File: SpatialGDK/Source/SpatialGDK/Private/Interop/ActorSystem.cpp [hunk 5]
```diff
@@ -791,6 +791,11 @@ void ActorSystem::EntityRemoved(const Worker_EntityId EntityId)
 	ActorDataStore.Remove(EntityId);
 }
 
+void ActorSystem::AddEntityToRefreshDormancy(Worker_EntityId EntityId, bool bMakeDormant)
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/ActorSystem.cpp:794]:
```suggestion
void ActorSystem::RefreshActorDormancyOnEntityCreation(Worker_EntityId EntityId, bool bMakeDormant)
```

File: SpatialGDK/Source/SpatialGDK/Private/Interop/ActorSystem.cpp [hunk 6]
```diff
@@ -791,6 +791,11 @@ void ActorSystem::EntityRemoved(const Worker_EntityId EntityId)
...
+{
+	EntitiesMapToRefreshDormancy.Emplace(EntityId, bMakeDormant);
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/ActorSystem.cpp:796]: similar change to this variable name

File: SpatialGDK/Source/SpatialGDK/Private/Interop/ActorSystem.cpp [hunk 7]
```diff
@@ -2493,6 +2498,13 @@ void ActorSystem::OnEntityCreated(const Worker_CreateEntityResponseOp& Op, FSpat
 		check(ClientSystemEntityId != SpatialConstants::INVALID_ENTITY_ID);
 		CommandsHandler.ClaimPartition(NetDriver->Connection->GetCoordinator(), ClientSystemEntityId, Op.entity_id);
 	}
+
+	if (EntitiesMapToRefreshDormancy.Contains(EntityId))
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/ActorSystem.cpp:2502]: Additional check that this was an actual successful creation, ie. check status_code == WORKER_STATUS_CODE_SUCCESS

File: SpatialGDK/Source/SpatialGDK/Public/Interop/ActorSystem.h [hunk 8]
```diff
@@ -132,6 +134,7 @@ class ActorSystem
 
 	void EntityAdded(Worker_EntityId EntityId);
 	void EntityRemoved(Worker_EntityId EntityId);
+
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Public/Interop/ActorSystem.h:137]:
```suggestion
```

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/DormancyTests/RefreshActorDormancyTest.cpp [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+	DOREPLIFETIME(ThisClass, AwakeToDormantActor);
+}
+
+void ARefreshActorDormancyTest::PrepareTest()
+{
+	Super::PrepareTest();
+
+	// Step 1 - Spawn dormancy actor which is non dormant and change NetDormancy to DORM_DormantAll after a tick
+	AddStep(TEXT("ServerSpawnNonDormatActor"), FWorkerDefinition::Server(1), nullptr, [this]() {
+		DormantToAwakeActor = SpawnActor<ARefreshActorDormancyTestActor>(FActorSpawnParameters(), /*bRegisterAsAutoDestroy*/ true);
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/DormancyTests/RefreshActorDormancyTest.cpp:30]:
```suggestion
		DormantToAwakeActor = SpawnActor<ARefreshActorDormancyTestActor>(FActorSpawnParameters());
```

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/DormancyTests/RefreshActorDormancyTest.cpp [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+		AwakeToDormantActor = SpawnActor<ARefreshActorDormancyTestActor>(FActorSpawnParameters(), /*bRegisterAsAutoDestroy*/ true);
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/DormancyTests/RefreshActorDormancyTest.cpp:31]:
```suggestion
		AwakeToDormantActor = SpawnActor<ARefreshActorDormancyTestActor>(FActorSpawnParameters());
```

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/DormancyTests/RefreshActorDormancyTest.cpp [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+		DormantToAwakeActor->SetupForTest(true);
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/DormancyTests/RefreshActorDormancyTest.cpp:32]: This function as a little obtuse to anyone reading the code. Perhaps change it to something like `SetInitiallyDormant`

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/DormancyTests/RefreshActorDormancyTestActor.cpp [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+
+#include "RefreshActorDormancyTestActor.h"
+
+#include "Engine/Classes/Components/StaticMeshComponent.h"
+#include "Engine/Classes/Materials/Material.h"
+#include "Net/UnrealNetwork.h"
+
+ARefreshActorDormancyTestActor::ARefreshActorDormancyTestActor()
+{
+	NetDormancy = DORM_Initial;
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/DormancyTests/RefreshActorDormancyTestActor.cpp:11]:
```suggestion
```

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/DormancyTests/RefreshActorDormancyTestActor.cpp [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+}
+
+bool ARefreshActorDormancyTestActor::ReplicateSubobjects(UActorChannel* Channel, FOutBunch* Bunch, FReplicationFlags* RepFlags)
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/DormancyTests/RefreshActorDormancyTestActor.cpp:14]: nit: recommend changing this to PreReplication as it's slightly simpler and more obvious your intention

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/DormancyTests/RefreshActorDormancyTestActor.cpp [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+{
+	if (bSetupForTest && bIsSetDormancyComplete)
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/DormancyTests/RefreshActorDormancyTestActor.cpp:16]:
```suggestion
	if (PostTickFlushDelegateHandle.IsValid() && bIsSetDormancyComplete)
```

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/DormancyTests/RefreshActorDormancyTestActor.cpp [hunk 15]
```diff
@@ -0,0 +1,46 @@
...
+	{
+		GetWorld()->OnPostTickFlush().Remove(PostTickFlushDelegateHandle);
+		bIsSetDormancyComplete = false;
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/DormancyTests/RefreshActorDormancyTestActor.cpp:19]:
```suggestion
```

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/DormancyTests/RefreshActorDormancyTest.cpp [hunk 16]
```diff
@@ -0,0 +1,45 @@
...
+		AwakeToDormantActor = SpawnActor<ARefreshActorDormancyTestActor>(FActorSpawnParameters(), true);
+		DormantToAwakeActor->SetInitiallyDormant(true);
+		AwakeToDormantActor->SetInitiallyDormant(false);
+		FinishStep();
+	});
+
+	// Step 2 - Client check NetDormancy is DORM_Awake for DormantToAwakeActor and DORM_DormantAll for AwakeToDormantActor
+	AddStep(("ClientAssertDormancyTestStates"), FWorkerDefinition::AllClients, nullptr, nullptr,
+			[this](float DeltaTime) {
+				RequireEqual_Int(DormantToAwakeActor->NetDormancy, DORM_Awake, TEXT("DormantToAwakeActor is Awake"));
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/DormancyTests/RefreshActorDormancyTest.cpp:40]: Should add a nullptr check for both of these incase the client ticks before they have received the replicated actors

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp [hunk 17]
```diff
@@ -3040,6 +3048,20 @@ void USpatialNetDriver::RefreshActorDormancy(AActor* Actor, bool bMakeDormant)
...
+	if (!Connection->GetCoordinator().HasEntity(EntityId))
+	{
+		if (ActorSystem->IsCreateEntityRequestInFlight(EntityId))
+		{
+			ActorSystem->RefreshActorDormancyOnEntityCreation(EntityId, bMakeDormant);
+			return;
+		}
+		else
+		{
+			UE_LOG(LogSpatialOSNetDriver, Verbose, TEXT("Entity should be checked out for actor (%s)"), *Actor->GetName());
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp:3060]:
```suggestion
			UE_LOG(LogSpatialOSNetDriver, Verbose, TEXT("Unable to flush dormancy on actor (%s), entity (%lld) not in view"), *Actor->GetName(), EntityId);
```

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp [hunk 18]
```diff
@@ -3040,6 +3048,20 @@ void USpatialNetDriver::RefreshActorDormancy(AActor* Actor, bool bMakeDormant)
 		return;
 	}
 
+	if (!Connection->GetCoordinator().HasEntity(EntityId))
+	{
+		if (ActorSystem->IsCreateEntityRequestInFlight(EntityId))
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp:3053]:
```suggestion
		if (ActorSystem != nullptr && ActorSystem->IsCreateEntityRequestInFlight(EntityId))
```

**Review by m-samiec — APPROVED** [review]
Nice work!
